### DISCUSSION TITLE
more efficient process isolation

### DISCUSSION THREAD

**Discussion by staabm** [infection/infection#1952] [Ideas]
I did not yet deep dive into infection or its implementation details, therefore take this with a grain of salt.

If I get it right all/most? tests in infection need to run in isolation to make super sure tests and mutants don't have side-effects on each other. as I am exploring ways to [make phpunit test isolation more efficient](https://github.com/sebastianbergmann/phpunit/issues/5749), I though about sharing my findings here... maybe infection can also do similar things or take inspiration from it to make the internals more efficient

----

array based command notation should be used when possible to reduce process creation overhead. see https://github.com/sebastianbergmann/phpunit/pull/5766

----

in case infection really needs to spawn new processes for every test, it might make sense to use `pcntl_fork` instead of a worker process as it seems to [incur 6-7x less process creation overhead](https://github.com/sebastianbergmann/phpunit/pull/5751#issuecomment-2002582219)

----

I am open to help and/or implement above ideas in case they seem useful/acceptable

I am not yet finished with phpunit, so maybe I can share more findings in the future.

**Comment by theofidry** [comment]
>array based command notation should be used when possible to reduce process creation overhead. see https://github.com/sebastianbergmann/phpunit/pull/5766

🤔 any idea what is the cause of this? It's quite surprising (didn't seem to find the answer from having a glance at the conversations)

>in case infection really needs to spawn new processes for every test, it might make sense to use pcntl_fork instead of a worker process as it seems to https://github.com/sebastianbergmann/phpunit/pull/5751#issuecomment-2002582219

Wouldn't there be an abstraction layer that would allow to easily switch from a driver to another? I don't remember the state of things but I vaguely remember using some Amphp library for this at one point

**Reply by staabm** [reply]
> any idea what is the cause of this?

see https://github.com/symfony/symfony/issues/43162#issuecomment-1787670224 - when using array notation `proc_open` will not open a intermediate shell and instead talks to the process directly. the newer the php version beeing used, the faster this will be.

> Wouldn't there be an abstraction layer that would allow to easily switch from a driver to another?

amphp and reactphp process components are meant for worker pools - this means processes are re-used and global state might leak from one worker into another. I did not yet find a process abstraction which supports `pcntl_fork` (which is not supported on windows btw).  `pcntl_fork` only really makes sense when the process beeing forked is used for a one-time job and killed afterwards. additionally it only makes sense if we are fine with the forked process to own all constants/classes etc. the main process already defined before forking.

**Reply by theofidry** [reply]
FWIW I think it would be interesting to have a dedicated package for that pcntl fork. I use the process approach too in https://github.com/webmozarts/console-parallelization because you want the process to be killed at the end of one job. But it's been a while, if you want we can do a quick Discord call one of those days to have a brief look at this lib / Infection / PHPUnit. I would hope Amphp or Reactphp would cover this case but maybe it's completely out of scope for them.

**Reply by staabm** [reply]
I will work thru fork based process isolation in phpunit with https://github.com/sebastianbergmann/phpunit/pull/5751

after this landed and I have a better idea about all the involved problems I will see to get it into reactphp/amphp etc
